internal/pkg/client/patch/patch.go

```diff
@@ -0,0 +1,210 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package patch
+
+import (
+	"fmt"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/jsonmergepatch"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// Patch object
+type Patch struct {
+	Type types.PatchType
+	Data []byte
+}
+
+var metadataAccessor = meta.NewAccessor()
+
+// GetClientSideApplyPatch updates a resource using the same approach as running `kubectl apply`.
+// The implementation here has been mostly extracted from the apply command: k8s.io/kubernetes/pkg/kubectl/cmd/apply.go
+func GetClientSideApplyPatch(current, desired runtime.Object) (Patch, error) {
+	var patch Patch
+	// Serialize the current configuration of the object.
+	currentb, err := runtime.Encode(unstructured.UnstructuredJSONScheme, current)
+	if err != nil {
+		return patch, fmt.Errorf("could not serialize current %v", current)
+	}
+	// Retrieve the last applied configuration of the object from the annotation.
+	last, err := GetLastApplied(current)
+	if err != nil {
+		return patch, fmt.Errorf("could not retrieve original configuration from %v", desired)
+	}
+
+	// Serialize the modified configuration of the object, populating the last applied annotation as well.
+	modified, err := GetModifiedConfiguration(desired, true)
+	if err != nil {
+		return patch, fmt.Errorf("could not serialize intended configuration from %v", desired)
+	}
+	gvk := desired.GetObjectKind().GroupVersionKind()
+
+	versionedObject, err := Scheme.New(gvk)
+	_, unversioned := Scheme.IsUnversioned(desired)
+	switch {
+	case runtime.IsNotRegisteredError(err) || unversioned:
+		preconditions := []mergepatch.PreconditionFunc{
+			mergepatch.RequireKeyUnchanged("apiVersion"),
+			mergepatch.RequireKeyUnchanged("kind"),
+			mergepatch.RequireMetadataKeyUnchanged("name"),
+		}
+		patch.Data, err = jsonmergepatch.CreateThreeWayJSONMergePatch(last, modified, currentb, preconditions...)
+		patch.Type = types.MergePatchType
+		if err != nil {
+			name, err := metadataAccessor.Name(desired)
+			if mergepatch.IsPreconditionFailed(err) {
+				return patch, fmt.Errorf("at least one of apiVersion, kind and name was changed for %s/%s", gvk, name)
+			}
+			return patch, err
+		}
+	case err != nil:
```

Comment:
I'd put this one last, it is kinda like the default case

Reply:
done.